tsl/src/data_node.c

```diff
@@ -86,8 +86,22 @@ get_database_info(Oid dbid, DbInfo *database)
 	dbrecord = (Form_pg_database) GETSTRUCT(dbtuple);
 
 	database->encoding = dbrecord->encoding;
-	database->collation = dbrecord->datcollate;
-	database->chartype = dbrecord->datctype;
+
+#if PG15_GE
+	Datum datum;
+	bool isnull;
+
+	datum = SysCacheGetAttr(DATABASEOID, dbtuple, Anum_pg_database_datcollate, &isnull);
+	Assert(!isnull);
+	database->collation = TextDatumGetCString(datum);
+
+	datum = SysCacheGetAttr(DATABASEOID, dbtuple, Anum_pg_database_datctype, &isnull);
+	Assert(!isnull);
+	database->chartype = TextDatumGetCString(datum);
```

Comment:
You have already fetched the `dbrecord` from `pg_database` above. It is not clear why you need to fetch it again. Both fields `datcollate` and `datctype` are available in `FormData_pg_database` as `text` fields, so this is surprising. Especially since you're not using `SysCacheGetAttr` in the pre-PG15 code.

This is how it is done in PostgreSQL, but could you add a comment on why this is necessary?

Reply:
/*
   * SysCacheGetAttr
   *
   *    Given a tuple previously fetched by SearchSysCache(),
   *    extract a specific attribute.
   *
   * This is equivalent to using heap_getattr() on a tuple fetched
   * from a non-cached relation.  Usually, this is only used for attributes
   * that could be NULL or variable length; the fixed-size attributes in
   * a system table are accessed just by mapping the tuple onto the C struct
   * declarations from include/catalog/.
   */
Since the field changes from fixed-length to varlen we need to go through SysCacheGetAttr